code/4_data_pipeline/3_dataset_C02.py

```diff
@@ -0,0 +1,145 @@
+
+    # Extract pivot_id, landsat_version, year, month, and day from the new filename format
+    parts = tif_filename.split('_')
+    pivot_id = parts[1]
+    # landsat_name = parts[2]  # OLD line, do not remove, just comment
+    landsat_version = parts[2]   # Renamed to match Step 2
+    year, month, day = parts[3].split('-')
+    day = day.split('.')[0]
+
+    # Build the same 5-element key used in Step 2
+    key = (pivot_id, landsat_version, year, month, day)
+
+    if key not in image_to_annotation:
+        continue
+
+    annotation = image_to_annotation[key]
+    tif_file_path = f'{tif_folder}/{tif_filename}'
+    with rasterio.open(tif_file_path) as src:
+        band_data = ['N/A'] * 19  # Ensure you have 19 bands to read
+
+        # Get the GeoJSON filename
+        geojson_filename = f'{tif_filename.split(".tif")[0]}.geojson'
+        geojson_file_path = os.path.join(geojson_folder, geojson_filename)
+
+        # Open the GeoJSON file
+        with open(geojson_file_path, 'r') as geojson_file:
+            geojson_data = json.load(geojson_file)
+
+        bottom_left = geojson_data['features'][0]['geometry']['coordinates'][0][0]
+        top_right = geojson_data['features'][0]['geometry']['coordinates'][0][2]
+
+        # Extract annotation data
+        annotation_list = annotation.get('annotations', [])
+        if annotation_list:
+            for ann_item in annotation_list:
+                for result in ann_item.get('result', []):
+                    if result['type'] == 'keypointlabels':
+                        x_percent = result['value']['x']
+                        y_percent = result['value']['y']
+                        label = result['value']['keypointlabels'][0]
+
+                        # Convert percentage to pixel coordinates
+                        x_pixel = int(x_percent * src.width / 100)
+                        y_pixel = int(y_percent * src.height / 100)
+                        x = bottom_left[0] + (top_right[0] - bottom_left[0]) * (x_percent / 100)
+                        y = bottom_left[1] + (top_right[1] - bottom_left[1]) * (y_percent / 100)
+
+                        # Extract the first 19 bands from the TIFF image
+                        for i in range(19):  # Assuming you want the first 19 bands
+                            band_index = i + 1  # Band indices in rasterio are 1-based
+                            band = src.read(band_index, window=((y_pixel, y_pixel+1), (x_pixel, x_pixel+1)))
+                            if band.size > 0:
+                                band_data[i] = band[0, 0]
+
+                        # Append data to list
+                        tif_id = tif_name_to_id[key]
+                        row = [
+                            tif_id, tif_filename, landsat_version, year, month, day,
+                            x_percent, y_percent, label, x, y
+                        ] + band_data
+                        data.append(row)
+
+# Step 5: Write data to Excel file
```

Comment:
saving as excel is a bit unconventional -- why not just a csv?

Reply:
It's not a big deal I mean this works it's just not usual. Next time I would say use a csv.